### ISSUE TITLE
Excessive memory usage caused by Samplers storing lists of indices

### ISSUE THREAD

**Issue by almson** [pytorch/pytorch#50201]
## 🐛 Bug

Samplers use excessive memory. Memory usage is unbounded and prevents use of data loaders which report a large length (>100 million examples).

For example, a moderately-sized audio dataset of 500 hours at 100 fps contains 180 million timesteps and reports this as its length. This leads to ~100GB of memory usage during multi-GPU training.

Samplers shouldn't use memory to store lists. They should use a stateless random permutation algorithm.

**Comment by mattivi** [comment]
Hi @almson, I have implemented the code you have in the PR. I am using python 3.7 so I the isqrt function is as below.  However, RandomSample appears to be slow. Do you have any stats that compare this solution of RandomSampler with SequentialSampler for dataset >100M rows?
```
def isqrt(n):
    """
    Return the integer part of the square root of the input.
    """
    n = operator.index(n)
    if n < 0:
        raise ValueError("isqrt() argument must be nonnegative")
    if n == 0:
        return 0
    c = (n.bit_length() - 1) // 2
    a = 1
    d = 0
    for s in reversed(range(c.bit_length())):
        # Loop invariant: (a-1)**2 < (n >> 2*(c - d)) < (a+1)**2
        e = d
        d = c >> s
        a = (a << d - e - 1) + (n >> 2*c - e - d + 1) // a
    return a - (a*a > n)
```

**Comment by almson** [comment]
@mattivi I think you can replace isqrt() with just int(sqrt()), you don't need to reimplement the entire function. I haven't done extensive benchmarking. Did you profile to see which part is slow? Each iteration just does a little math and should be very fast.

**Comment by leokraft** [comment]
Hi,
just want to report that this might still be an ongoing issue.

We're currently trying to train a model on a large dataset (>60m samples).
However, when training we notice that our CPU memory is steadily increasing.

This also happens when we're using a dummy Dataset i.e.

```
class DummyDataset(torch.utils.data.Dataset):

    def __init__(self):
        super().__init__()

    def __len__(self):
        return 60_000_000

    def __getitem__(self, idx: int):
        return torch.tensor([0, 1]), torch.tensor([2])
```

We're using `num_workers > 0` and for each worker there seems to be an additional ~1.5 GB of RAM consumption.

Our assumption is, that this happens due to `shuffle=True` in our DataLoader.
The RandomSampler in the DataLoader permutates the indices for sampling and converts them into a list of `int`s. ([here](https://github.com/pytorch/pytorch/blob/71ad1005f66c9a53a2fe28d24b95c4e828aa944e/torch/utils/data/sampler.py#L132))

For our example and on our machine (where `sys.getsizeof(some_int)==28`) the size of this list would then be around:

`28B * 60 000 000 = 1.68GB`

Which would be copied for each worker and explain the memory consumption.

[Original post](https://discuss.pytorch.org/t/randomsampler-uses-excessive-amounts-of-memory-for-indices-list-when-shuffling)
